File: Melina-Aguilar/JavaScript/Clase 3/PasoPorReferencia.js
/* Paso por referencia: a la funcion le pasamos la referencia hexadecimal de el espacio de
 memoria donde estaba alojado el objeto. Cuando creamos el objeto se creo un espacio de memoria
 y al pasarle ese objeto en el llamado de la funcion (como un parametro) le estamos pasando
 directamente esa valor hexadecimal de la memoria y de esta forma dentro de la funcion a traves
 de la variable 'persona1' hemos modificado esos valores y luego esa variable se destruya al
 finalizar la funcion pero la modificiacion quedo grabada dentro del objeto 'persona'.
 */

// Definimos un objeto
const persona = {
    nombre: 'Juan',
    apellido: 'Lopez'
}

console.log(persona);

function cambiarValorObjeto(persona1){
    persona1.nombre = 'Ignacio';
    persona1.apellido = 'Perez';
}

cambiarValorObjeto(persona); // La variable 'persona1' al finalizar la funcion se destruye
                             // pero las modificaciones afectaran permanentemente al objeto 'persona'.
console.log(persona);

File: Java/Leccion2/LeccionClaseObjetos/src/PasoPorReferencia/PasoPorReferencia.java
package PasoPorReferencia;

import Persona.Persona;

public class PasoPorReferencia {
    public static void main(String[] args) {
        Persona persona1 = new Persona();
        persona1.nombre = "Ester";
        System.out.println("Nombre de la persona = " + persona1.nombre);

        cambiarValor(persona1);
        System.out.println("El cambio que hicimos en el nombre es: " + persona1.nombre);
        persona1 = cambiarElValor(persona1);
        System.out.println("El cambio que hicimos en el nombre es: " + persona1.nombre);
        persona1 = cambiarElValor(persona1);
        Persona persona2 = null;
        persona2  = cambiarElValor(persona2);
        System.out.println("El nuevo valor del objeto es: "+persona2.nombre);

    }

    public static void cambiarValor(Persona persona) {
        // Paso por referencia
        persona.nombre = "Maria";
    }


    public static Persona cambiarElValor(Persona persona) {
        if (persona == null) {
            System.out.println("Valor de persona es invalido :NULL");
            return null;
        } else {


            persona.nombre = "Monica";
            return persona;
        }
    }

}